python/pyarrow/_substrait.pyx

```diff
@@ -27,7 +27,7 @@ from pyarrow.includes.libarrow cimport *
 from pyarrow.includes.libarrow_substrait cimport *
 
 
-cdef CDeclaration _create_named_table_provider(dict named_args, const std_vector[c_string]& names):
+cdef CDeclaration _create_named_table_provider(dict named_args, const std_vector[c_string]& names, const CSchema& schema):
```

Comment:
I suppose we would eventually want to pass this schema on to the python function correct?

Something like:

```
    c_schema_sp = shared_ptr[CSchema](schema)
    py_schema = wrap_schema(c_schema_sp)

    py_table = named_args["provider"](py_names, py_schema)
```

Do you not need this for your use case?

Reply:
Yeah the python side support for this is left out but we only use named table provider via the C++ API.

I didn't do the python side because (1) we don't use it and (2) I am not very familiar with the python C extension. (Took me a while to figure the current C extension changes just to get it working).

Reply:
I can spend a bit time to see if I can get this working on the Python side. Looks like your example code is simple enough.

Reply:
Hmm, any quick ideas on the error?

```
    py_names = []
    for i in range(names.size()):
        c_name = names[i]
        py_names.append(frombytes(c_name))
    c_schema_sp = shared_ptr[CSchema](schema)
                                    ^
------------------------------------------------------------

pyarrow/_substrait.pyx:42:37: Cannot convert 'shared_ptr[CSchema]' to Python object
gmake[2]: *** [CMakeFiles/_substrait_pyx.dir/build.make:71: CMakeFiles/_substrait_pyx] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:785: CMakeFiles/_substrait_pyx.dir/all] Error 2
```

Reply:
Hmm I might be importing `wrap_schema` from the wrong place:

```
from pyarrow.lib import frombytes, tobytes, wrap_schema
```

I think it's complaining about 
```
py_schema = wrap_schema(c_schema_sp)
```
(error trace is weird on the wrong line)

Reply:
Ok I managed to get extesion to compile with

```
c_in_schema = make_shared[CSchema](move(schema))
```

Let me fix  the python test next

Reply:
Hmm, the first error was probably that I didn't point out that you would need to add `c_schema_sp` to your `cdef` section (so cython was recognizing `c_schema_sp` as a python object and not a C++ object).

Reply:
Actually - hold off - I think I didn't handle ownership correctly - should be probably doing a copy instead of move here. Let me see if I can fix.

Reply:
Ok I pushed another revision - I think this is correct now